File: lib/services/enpit_sirves.dart
import 'package:flutter/foundation.dart';
import 'dart:developer' as developer;

class EnpitSirves extends ChangeNotifier {

  final StringBuffer _inputText = StringBuffer();
  bool _isLetterMode = true;
  bool _shouldCapitalize = false;
  bool _isTextFieldFocused = false;

  String get inputText => _inputText.toString();
  bool get isLetterMode => _isLetterMode;
  bool get isTextFieldFocused => _isTextFieldFocused;

  void addCharacter(String char) {
    developer.log('EnpitSirves: addCharacter called with: $char');
    if (char == '⌫') {
      deleteLeft();
      developer.log('EnpitSirves: Deleting left via addCharacter.');
      return;
    }

    String finalChar = _shouldCapitalize ? char.toUpperCase() : char;
    _inputText.write(finalChar);
    _shouldCapitalize = false;
    developer.log('EnpitSirves: _inputText updated to: "${_inputText.toString()}"');
    notifyListeners();
    developer.log('EnpitSirves: notifyListeners called.');
  }

  void deleteLeft() {
    if (_inputText.isNotEmpty) {
      String currentText = _inputText.toString();
      _inputText.clear();
      _inputText.write(currentText.substring(0, currentText.length - 1));
      notifyListeners();
    }
  }

  void deleteWord() {
    if (_inputText.isEmpty) return;

    String currentText = _inputText.toString();
    String trimmedText = currentText.trimRight();
    int lastSpaceIndex = trimmedText.lastIndexOf(' ');

    _inputText.clear();
    if (lastSpaceIndex != -1) {
      _inputText.write(trimmedText.substring(0, lastSpaceIndex));
    }
    else {
      _inputText.write('');
    }
    notifyListeners();
  }

  void deleteCharacters(int count) {
    String currentText = _inputText.toString();
    if (currentText.isNotEmpty && currentText.length >= count) {
      _inputText.clear();
      _inputText.write(currentText.substring(0, currentText.length - count));
      notifyListeners();
    } else if (currentText.isNotEmpty && currentText.length < count) {
      _inputText.clear();
      notifyListeners();
    }
  }

  void toggleMode() {
    _isLetterMode = !_isLetterMode;
    notifyListeners();
  }

  void setCapitalize() {
    _shouldCapitalize = true;
  }

  void setTextFieldFocus(bool focused) {
    _isTextFieldFocused = focused;
    notifyListeners();
  }

  String getDisplayText(int displayLength) {
    if (_inputText.isEmpty) return ' ' * displayLength;
    String currentText = _inputText.toString();
    if (currentText.length >= displayLength) {
      return currentText.substring(currentText.length - displayLength);
    } else {
      return currentText.padLeft(displayLength, ' ');
    }
  }

  void clearText() {
    _inputText.clear();
    developer.log(
        'InputService: _inputText cleared. New value: "${_inputText.toString()}"');
    notifyListeners();
  }
}

File: kepad/src/commonMain/kotlin/sirvesez/EnpitSirves.kt
package sirvesez

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update

/**
 * A service to manage the input state of the keyboard.
 *
 * This class is a translation of a Dart ChangeNotifier, using StateFlow for reactive
 * state management in Kotlin. In a real application, consider using a
 * proper logging library instead of println.
 */
class EnpitSirves {

    private val _inputText = MutableStateFlow("")
    val inputText = _inputText.asStateFlow()

    private val _isLetterMode = MutableStateFlow(true)
    val isLetterMode = _isLetterMode.asStateFlow()

    private var _shouldCapitalize = false

    private val _isTextFieldFocused = MutableStateFlow(false)
    val isTextFieldFocused = _isTextFieldFocused.asStateFlow()

    fun addCharacter(char: String) {
        println("EnpitSirves: addCharacter called with: $char")
        if (char == "⌫") {
            deleteLeft()
            println("EnpitSirves: Deleting left via addCharacter.")
            return
        }

        val finalChar = if (_shouldCapitalize) char.uppercase() else char
        _inputText.update { it + finalChar }
        _shouldCapitalize = false
        println("EnpitSirves: _inputText updated to: \"${_inputText.value}\" ")
    }

    fun deleteLeft() {
        _inputText.update { it.dropLast(1) }
    }

    fun deleteWord() {
        if (_inputText.value.isEmpty()) return

        val currentText = _inputText.value
        val trimmedText = currentText.trimEnd()
        val lastSpaceIndex = trimmedText.lastIndexOf(' ')

        _inputText.value = if (lastSpaceIndex != -1) {
            trimmedText.substring(0, lastSpaceIndex)
        } else {
            ""
        }
    }

    fun deleteCharacters(count: Int) {
        val currentText = _inputText.value
        if (currentText.length >= count) {
            _inputText.value = currentText.dropLast(count)
        } else {
            _inputText.value = ""
        }
    }

    fun toggleMode() {
        _isLetterMode.update { !it }
    }

    fun setCapitalize() {
        _shouldCapitalize = true
    }

    fun setTextFieldFocus(focused: Boolean) {
        _isTextFieldFocused.value = focused
    }

    fun getDisplayText(displayLength: Int): String {
        val currentText = _inputText.value
        return when {
            currentText.isEmpty() -> " ".repeat(displayLength)
            currentText.length >= displayLength -> currentText.takeLast(displayLength)
            else -> currentText.padStart(displayLength, ' ')
        }
    }

    fun clearText() {
        _inputText.value = ""
        println("InputService: _inputText cleared. New value: \"${_inputText.value}\" ")
    }
}
